cmd/kubetoken/validator.go

```diff
@@ -0,0 +1,429 @@
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"sync"
+	"time"
+
+	oktetoLog "github.com/okteto/okteto/pkg/log"
+	"github.com/okteto/okteto/pkg/okteto"
+	"github.com/okteto/okteto/pkg/types"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+)
+
+const (
+	tokenRequestKind = "TokenRequest"
+)
+
+var (
+	errEmptyNamespace           = errors.New("namespace cannot be empty")
+	errEmptyContext             = errors.New("context name cannot be empty")
+	errTokenRequestNotSupported = errors.New("kubernetes cluster does not support TokenRequest")
+
+	valdationTimeout = 3 * time.Second
+)
+
+// validator is the interface that wraps the Validate method.
+type validator interface {
+	Validate(ctx context.Context) error
+}
+
+type preReqCfg struct {
+	ctxName string
+	ns      string
+
+	k8sClientProvider    okteto.K8sClientProvider
+	oktetoClientProvider types.OktetoClientProvider
+}
+
+type option func(*preReqCfg)
+
+func withCtxName(ctxName string) option {
+	return func(cfg *preReqCfg) {
+		cfg.ctxName = ctxName
+	}
+}
+
+func withNamespace(ns string) option {
+	return func(cfg *preReqCfg) {
+		cfg.ns = ns
+	}
+}
+
+func withK8sClientProvider(k8sClientProvider okteto.K8sClientProvider) option {
+	return func(cfg *preReqCfg) {
+		cfg.k8sClientProvider = k8sClientProvider
+	}
+}
+
+func withOktetoClientProvider(oktetoClientProvider types.OktetoClientProvider) option {
+	return func(cfg *preReqCfg) {
+		cfg.oktetoClientProvider = oktetoClientProvider
+	}
+}
+
+func defaultPreReqCfg() *preReqCfg {
+	return &preReqCfg{
+		k8sClientProvider:    okteto.NewK8sClientProvider(),
+		oktetoClientProvider: okteto.NewOktetoClientProvider(),
+	}
+}
+
+// preReqValidator validates that all the pre-reqs to execute the command are met
+type preReqValidator struct {
+	ctxName string
+	ns      string
+
+	k8sClientProvider    okteto.K8sClientProvider
+	oktetoClientProvider types.OktetoClientProvider
+}
+
+// newPreReqValidator returns a new preReqValidator
+func newPreReqValidator(opts ...option) *preReqValidator {
+	cfg := defaultPreReqCfg()
+	for _, opt := range opts {
+		opt(cfg)
+	}
+	return &preReqValidator{
+		ctxName:              cfg.ctxName,
+		ns:                   cfg.ns,
+		k8sClientProvider:    cfg.k8sClientProvider,
+		oktetoClientProvider: cfg.oktetoClientProvider,
+	}
+}
+
+// Validate validates that all the pre-reqs to execute the command are met
+func (v *preReqValidator) Validate(ctx context.Context) error {
+	oktetoLog.Info("validating pre-reqs")
+
+	ctx, cancel := context.WithTimeout(ctx, valdationTimeout)
+	defer cancel()
+
+	err := newCtxValidator(v.ctxName, v.k8sClientProvider).Validate(ctx)
+	if err != nil {
+		return fmt.Errorf("invalid context: %w", err)
+	}
+
+	var wg sync.WaitGroup
+	validators := []validator{
+		newNsValidator(v.ns, v.oktetoClientProvider),
+		newOktetoSupportValidator(ctx, v.ctxName, v.ns, v.k8sClientProvider, v.oktetoClientProvider),
```

Comment:
there is no dependency between ns validator and ctx validator I think. Could both run in parallel and run support validator once ctx validator is finished? even if ns validator is not finished. Something like:

````
ns validator | ctx validator | (if ctx validator is finished) okteto support validator
````

Reply:
there is. To create the okteto client we need the okteto url which is the name of the context. If the context doesn't exists it will fail